File: Android/BlueMessenger/app/src/main/java/com/omercankoc/bluemessenger/ChatClass.java
package com.omercankoc.bluemessenger;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothServerSocket;
import android.bluetooth.BluetoothSocket;
import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.Log;

import java.io.IOException;
import java.util.UUID;


public class ChatClass {
    private Context context;
    private final Handler handler;
    private BluetoothAdapter bluetoothAdapter;

    private ConnectThread connectThread;
    private AcceptThread acceptThread;

    private final UUID APP_UUID = UUID.fromString("fa87c0c0-afac-11de-8a39-0800200c9a66");
    private final String APP_NAME = "BlueMessanger";

    public static final int STATE_NONE = 0;
    public static final int STATE_LISTEN = 1;
    public static final int STATE_CONNECTING = 2;
    public static final int STATE_CONNECTED = 3;

    private int state;

    public ChatClass(Context context, Handler handler){
        this.context = context;
        this.handler = handler;

        state = STATE_NONE;
        bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
    }

    public int getData(){
        return state;
    }

    public synchronized void setState(int state){
        this.state = state;
        handler.obtainMessage(MainActivity.MESSAGE_STATE_CHANGED,state,-1).sendToTarget();
    }

    private synchronized void start(){
        if(connectThread != null){
            connectThread.cancel();
            connectThread = null;
        }

        if(acceptThread == null){
            acceptThread = new AcceptThread();
            acceptThread.start();
        }
        setState(STATE_LISTEN);
    }

    public synchronized void stop(){
        if(connectThread != null){
            connectThread.cancel();
            connectThread = null;
        }

        if(acceptThread == null){
            acceptThread.cancel();
            acceptThread = null;
        }
        setState(STATE_NONE);
    }

    public void connect(BluetoothDevice bluetoothDevice){
        if(state == STATE_CONNECTING){
            connectThread.cancel();
            connectThread = null;
        }

        connectThread = new ConnectThread(bluetoothDevice);
        connectThread.start();

        setState(STATE_CONNECTING);
    }

    /*

     */
    private class AcceptThread extends Thread{
        private BluetoothServerSocket serverSocket;

        public AcceptThread(){
            BluetoothServerSocket tmp = null;
            try {
                tmp = bluetoothAdapter.listenUsingRfcommWithServiceRecord(APP_NAME,APP_UUID);
            } catch (IOException e){
                Log.e("Accept -> Constructor",e.toString());
            }

            serverSocket = tmp;
        }

        @Override
        public void run() {
            BluetoothSocket bluetoothSocket = null;
            try{
                serverSocket.accept();
            } catch (IOException e){
                Log.e("Accept -> Run",e.toString());
                try {
                    serverSocket.close();
                } catch (IOException e1){
                    Log.e("Accept -> Close",e.toString());
                }
            }

            if(bluetoothSocket != null){
                switch (state){
                    case STATE_LISTEN:
                    case STATE_CONNECTING:
                        connect(bluetoothSocket.getRemoteDevice());
                        break;
                    case STATE_NONE:
                    case STATE_CONNECTED:
                        try {
                            bluetoothSocket.close();
                        } catch (IOException e){
                            Log.e("Accept -> Close Socket",e.toString());
                        }
                }
            }
        }

        public void cancel(){
            try {
                serverSocket.close();
            } catch (IOException e){
                Log.e("Accept -> Close Server",e.toString());
            }
        }
    }

    /*

     */
    private class ConnectThread extends Thread{
        private final BluetoothSocket bluetoothSocket;
        private final BluetoothDevice bluetoothDevice;

        public ConnectThread(BluetoothDevice bluetoothDevice){
            this.bluetoothDevice = bluetoothDevice;
            BluetoothSocket tmp = null;
            try {
                tmp = bluetoothDevice.createRfcommSocketToServiceRecord(APP_UUID);
            } catch(IOException e){
                Log.e("Connect -> Constructor",e.toString());
            }
            bluetoothSocket = tmp;
        }

        public void run(){
            try {
                bluetoothSocket.connect();
            } catch (IOException e){
                Log.e("Connect -> Run",e.toString());
                try {
                    bluetoothSocket.close();
                } catch (IOException e1){
                    Log.e("Connect -> Close Socket",e1.toString());
                }
                connectionFailed();
                return;
            }

            synchronized (ChatClass.this){
                connectThread = null;
            }

            connected(bluetoothDevice);
        }

        public void cancel(){
            try {
                bluetoothSocket.close();
            } catch (IOException e){
                Log.e("Connect -> Cancel",e.toString());
            }
        }
    }

    /*

     */
    private synchronized void connectionFailed(){
        Message message = handler.obtainMessage(MainActivity.MESSAGE_TOAST);
        Bundle bundle = new Bundle();
        bundle.putString(MainActivity.TOAST,"Can not connected device!");
        message.setData(bundle);
        handler.sendMessage(message);

        ChatClass.this.start();
    }

    /*

     */
    private synchronized void connected(BluetoothDevice bluetoothDevice){
        if(connectThread != null){
            connectThread.cancel();
            connectThread = null;
        }

        Message message = handler.obtainMessage(MainActivity.MESSAGE_DEVICE_NAME);
        Bundle bundle = new Bundle();
        bundle.putString(MainActivity.DEVICE_NAME,bluetoothDevice.getName());
        message.setData(bundle);
        handler.sendMessage(message);

        setState(STATE_CONNECTED);
    }
}

File: Android/BlueMessenger/app/src/main/java/com/omercankoc/bluemessenger/ChatClass.kt
package com.omercankoc.bluemessenger

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothServerSocket
import android.bluetooth.BluetoothSocket
import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.util.Log
import com.omercankoc.bluemessenger.ChatClass
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.*

class ChatClass(private val context: Context, private val handler: Handler) {
    private val bluetoothAdapter: BluetoothAdapter
    private var connectThread: ConnectThread? = null
    private var acceptThread: AcceptThread? = null
    private var connectedThread: ConnectedThread? = null
    private val APP_UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB")
    private val APP_NAME = "BlueMessenger"
    private var state: Int
    fun getState(): Int {
        return state
    }

    @Synchronized
    fun setState(state: Int) {
        this.state = state
        handler.obtainMessage(MainActivity.MESSAGE_STATE_CHANGED, state, -1).sendToTarget()
    }

    @Synchronized
    private fun start() {
        if (connectThread != null) {
            connectThread!!.cancel()
            connectThread = null
        }
        if (acceptThread == null) {
            acceptThread = AcceptThread()
            acceptThread!!.start()
        }
        if (connectedThread != null) {
            connectedThread!!.cancel()
            connectedThread = null
        }
        setState(STATE_LISTEN)
    }

    @Synchronized
    fun stop() {
        if (connectThread != null) {
            connectThread!!.cancel()
            connectThread = null
        }
        if (acceptThread != null) {
            acceptThread!!.cancel()
            acceptThread = null
        }
        if (connectedThread != null) {
            connectedThread!!.cancel()
            connectedThread = null
        }
        setState(STATE_NONE)
    }

    fun connect(bluetoothDevice: BluetoothDevice) {
        if (state == STATE_CONNECTING) {
            connectThread!!.cancel()
            connectThread = null
        }
        connectThread = ConnectThread(bluetoothDevice)
        connectThread!!.start()
        if (connectedThread != null) {
            connectedThread!!.cancel()
            connectedThread = null
        }
        setState(STATE_CONNECTING)
    }

    fun write(buffer: ByteArray?) {
        var connThread: ConnectedThread?
        synchronized(this) {
            if (state != STATE_CONNECTED) {
                return
            }
            connThread = connectedThread
        }
        connThread!!.write(buffer)
    }

    /*

     */
    private inner class AcceptThread : Thread() {
        private val serverSocket: BluetoothServerSocket?
        override fun run() {
            var bluetoothSocket: BluetoothSocket? = null
            try {
                bluetoothSocket = serverSocket!!.accept()
            } catch (e: IOException) {
                Log.e("Accept -> Run", e.toString())
                try {
                    serverSocket!!.close()
                } catch (e1: IOException) {
                    Log.e("Accept -> Close", e.toString())
                }
            }
            if (bluetoothSocket != null) {
                when (state) {
                    STATE_LISTEN, STATE_CONNECTING -> connected(bluetoothSocket, bluetoothSocket.remoteDevice)
                    STATE_NONE, STATE_CONNECTED -> try {
                        bluetoothSocket.close()
                    } catch (e: IOException) {
                        Log.e("Accept -> Close Socket", e.toString())
                    }
                }
            }
        }

        fun cancel() {
            try {
                serverSocket!!.close()
            } catch (e: IOException) {
                Log.e("Accept -> Close Server", e.toString())
            }
        }

        init {
            var tmp: BluetoothServerSocket? = null
            try {
                tmp = bluetoothAdapter.listenUsingRfcommWithServiceRecord(APP_NAME, APP_UUID)
            } catch (e: IOException) {
                Log.e("Accept -> Constructor", e.toString())
            }
            serverSocket = tmp
        }
    }

    private inner class ConnectThread(private val bluetoothDevice: BluetoothDevice) : Thread() {
        private val bluetoothSocket: BluetoothSocket?
        override fun run() {
            try {
                bluetoothSocket!!.connect()
            } catch (e: IOException) {
                Log.e("Connect -> Run", e.toString())
                try {
                    bluetoothSocket!!.close()
                } catch (e1: IOException) {
                    Log.e("Connect -> Close Socket", e1.toString())
                }
                connectionFailed()
                return
            }
            synchronized(this@ChatClass) { connectThread = null }
            connected(bluetoothSocket, bluetoothDevice)
        }

        fun cancel() {
            try {
                bluetoothSocket!!.close()
            } catch (e: IOException) {
                Log.e("Connect -> Cancel", e.toString())
            }
        }

        init {
            var tmp: BluetoothSocket? = null
            try {
                tmp = bluetoothDevice.createRfcommSocketToServiceRecord(APP_UUID)
            } catch (e: IOException) {
                Log.e("Connect -> Constructor", e.toString())
            }
            bluetoothSocket = tmp
        }
    }

    private inner class ConnectedThread(private val socket: BluetoothSocket?) : Thread() {
        private val inputStream: InputStream?
        private val outputStream: OutputStream?
        override fun run() {
            val buffer = ByteArray(1024)
            val bytes: Int
            try {
                bytes = inputStream!!.read(buffer)
                handler.obtainMessage(MainActivity.MESSAGE_READ, bytes, -1, buffer).sendToTarget()
            } catch (e: IOException) {
                connectionLost()
            }
        }

        fun write(buffer: ByteArray?) {
            try {
                outputStream!!.write(buffer)
                handler.obtainMessage(MainActivity.MESSAGE_WRITE, -1, -1, buffer).sendToTarget()
            } catch (e: IOException) {
                Log.e("Do not write message!", e.toString())
            }
        }

        fun cancel() {
            try {
                socket!!.close()
            } catch (e: IOException) {
                Log.e("Do not cancel thread!", e.toString())
            }
        }

        init {
            var tmpIn: InputStream? = null
            var tmpOut: OutputStream? = null
            try {
                tmpIn = socket!!.inputStream
                tmpOut = socket.outputStream
            } catch (e: IOException) {
            }
            inputStream = tmpIn
            outputStream = tmpOut
        }
    }

    private fun connectionLost() {
        val message = handler.obtainMessage(MainActivity.MESSAGE_TOAST)
        val bundle = Bundle()
        bundle.putString(MainActivity.TOAST, "Connection Lost")
        message.data = bundle
        handler.sendMessage(message)
        start()
    }

    /*

     */
    @Synchronized
    private fun connectionFailed() {
        val message = handler.obtainMessage(MainActivity.MESSAGE_TOAST)
        val bundle = Bundle()
        bundle.putString(MainActivity.TOAST, "Can not connected device!")
        message.data = bundle
        handler.sendMessage(message)
        start()
    }

    /*

     */
    @Synchronized
    private fun connected(bluetoothSocket: BluetoothSocket?, bluetoothDevice: BluetoothDevice) {
        if (connectThread != null) {
            connectThread!!.cancel()
            connectThread = null
        }
        if (connectedThread != null) {
            connectedThread!!.cancel()
            connectedThread = null
        }
        connectedThread = ConnectedThread(bluetoothSocket)
        connectedThread!!.start()
        val message = handler.obtainMessage(MainActivity.MESSAGE_DEVICE_NAME)
        val bundle = Bundle()
        bundle.putString(MainActivity.DEVICE_NAME, bluetoothDevice.name)
        message.data = bundle
        handler.sendMessage(message)
        setState(STATE_CONNECTED)
    }

    companion object {
        const val STATE_NONE = 0
        const val STATE_LISTEN = 1
        const val STATE_CONNECTING = 2
        const val STATE_CONNECTED = 3
    }

    init {
        state = STATE_NONE
        bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
    }
}